tools/src/test/java/org/apache/kafka/tools/reassign/ReassignPartitionsIntegrationTest.java

```diff
@@ -0,0 +1,924 @@
+        // In this test case, broker 0 starts up first on the latest IBP and is typically
+        // elected as controller. The three remaining brokers start up on the older IBP.
+        // We want to ensure that reassignment can still complete through the ISR change
+        // notification path even though the controller expects `AlterPartition`.
+
+        // Override change notification settings so that test is not delayed by ISR
+        // change notification delay
+        ZkAlterPartitionManager.DefaultIsrPropagationConfig_$eq(new IsrChangePropagationConfig(500, 100, 500));
+
+        Map<String, String> oldIbpConfig = Collections.singletonMap(KafkaConfig.InterBrokerProtocolVersionProp(), IBP_2_7_IV1.version());
+        Map<Integer, Map<String, String>> brokerConfigOverrides = new HashMap<>();
+        brokerConfigOverrides.put(1, oldIbpConfig);
+        brokerConfigOverrides.put(2, oldIbpConfig);
+        brokerConfigOverrides.put(3, oldIbpConfig);
+
+        cluster = new ReassignPartitionsTestCluster(Collections.emptyMap(), brokerConfigOverrides);
+        cluster.setup();
+
+        executeAndVerifyReassignment();
+    }
+
+    private void executeAndVerifyReassignment() throws ExecutionException, InterruptedException, JsonProcessingException {
+        String assignment = "{\"version\":1,\"partitions\":" +
+            "[{\"topic\":\"foo\",\"partition\":0,\"replicas\":[0,1,3],\"log_dirs\":[\"any\",\"any\",\"any\"]}," +
+            "{\"topic\":\"bar\",\"partition\":0,\"replicas\":[3,2,0],\"log_dirs\":[\"any\",\"any\",\"any\"]}" +
+            "]}";
+
+        TopicPartition foo0 = new TopicPartition("foo", 0);
+        TopicPartition bar0 = new TopicPartition("bar", 0);
+
+        // Check that the assignment has not yet been started yet.
+        Map<TopicPartition, PartitionReassignmentState> initialAssignment = new HashMap<>();
+
+        initialAssignment.put(foo0, new PartitionReassignmentState(Arrays.asList(0, 1, 2), Arrays.asList(0, 1, 3), true));
+        initialAssignment.put(bar0, new PartitionReassignmentState(Arrays.asList(3, 2, 1), Arrays.asList(3, 2, 0), true));
+
+        waitForVerifyAssignment(cluster.adminClient, assignment, false,
+            new VerifyAssignmentResult(initialAssignment));
+
+        // Execute the assignment
+        runExecuteAssignment(cluster.adminClient, false, assignment, -1L, -1L);
+        assertEquals(unthrottledBrokerConfigs, describeBrokerLevelThrottles(unthrottledBrokerConfigs.keySet()));
+        Map<TopicPartition, PartitionReassignmentState> finalAssignment = new HashMap<>();
+        finalAssignment.put(foo0, new PartitionReassignmentState(Arrays.asList(0, 1, 3), Arrays.asList(0, 1, 3), true));
+        finalAssignment.put(bar0, new PartitionReassignmentState(Arrays.asList(3, 2, 0), Arrays.asList(3, 2, 0), true));
+
+        kafka.admin.ReassignPartitionsCommand.VerifyAssignmentResult verifyAssignmentResult = runVerifyAssignment(cluster.adminClient, assignment, false);
```

Comment:
can we import VerifyAssignmentResult to avoid this whole path?

Reply:
Actually, we can't.

Because, java counterpart of POJO classes already exists in trunk in the same package as test.
So to use scala version we have to use FQN.

[PartitionReassignmentState](https://github.com/apache/kafka/blob/trunk/tools/src/main/java/org/apache/kafka/tools/reassign/PartitionReassignmentState.java)

Reply:
Just for the record - FQN will gone when main PR #13247 will be merged. So only one version of POJO classes (java one) will exists in codebase.

Reply:
Ah, I guess I was thinking of Scala where we can alias import.